File: src/previous/day82/Trie.ts
class Node {
  public word?: string
  public children: Map<string, Node>
  constructor(word?: string) {
    this.word = word
    this.children = new Map()
  }
}

export default class Trie {
  root: Node
  constructor() {
    this.root = new Node()
  }

  insert(item: string): void {
    let curr = this.root
    for (let char of item) {
      if (!curr.children.has(char)) curr.children.set(char, new Node())
      curr = curr.children.get(char)!
    }

    curr.word = item
  }
  delete(item: string): void {
    recurse(this.root)

    function recurse(curr: Node, idx: number = 0): boolean {
      if (idx === item.length) {
        if (curr.word) delete curr.word
        return curr.children.size === 0
      }

      const char = item[idx]
      const child = curr.children.get(char)
      if (child && !recurse(child, idx + 1)) return false

      curr.children.delete(char)

      return curr.children.size === 0
    }
  }
  find(partial: string): string[] {
    let curr = this.root

    for (let char of partial) {
      if (!curr.children.has(char)) return []
      curr = curr.children.get(char)!
    }

    const out: string[] = []
    recurse(curr)
    return out

    function recurse(curr: Node): void {
      if (curr.word) out.push(curr.word)
      for (let child of curr.children.values()) {
        recurse(child)
      }
    }
  }
}

File: go/day01/Trie/Trie.go
package trie

import (
	"unicode"
)

type Node struct {
	children map[rune]*Node
	word     string
}

type Trie struct {
	root *Node
}

func MakeTrie() *Trie {
	root := &Node{
		children: make(map[rune]*Node),
	}
	return &Trie{
		root: root,
	}
}

func (t *Trie) Insert(word string) {
	curr := t.root

	for _, r := range word {
		char := unicode.ToLower(r)
		child, ok := curr.children[char]
		if !ok {
			curr.children[char] = &Node{
				children: make(map[rune]*Node),
			}
			child = curr.children[char]
		}

		curr = child
	}
	curr.word = word
}

func (t *Trie) Find(prefix string) []string {
	curr := t.root
	for _, r := range prefix {
		char := unicode.ToLower(r)
		child, ok := curr.children[char]
		if !ok {
			return []string{}
		}

		curr = child
	}

	var out []string
	var recurse func(curr *Node)
	recurse = func(curr *Node) {
		if curr.word != "" {
			out = append(out, curr.word)
		}
		for _, child := range curr.children {
			recurse(child)
		}
	}

	recurse(curr)
	return out
}

func (t *Trie) Delete(word string) {
	var recurse func(curr *Node, idx int) bool
	recurse = func(curr *Node, idx int) bool {
		if curr.word != "" {
			curr.word = ""
			return len(curr.children) == 0
		}

		char := unicode.ToLower(rune(word[idx]))
		child, ok := curr.children[char]
		if !ok {
			return false
		}

		if !recurse(child, idx+1) {
			return false
		}

		delete(curr.children, char)
		return len(curr.children) == 0
	}

	recurse(t.root, 0)
}
